### PR TITLE
Introduce MSVC CI support

### LINKED ISSUE #24: MSVC CI Support
It's important for Beman projects to support MSVC in continuous integration, especially since not all contributors will have immediate access to that platform.

This issue will be done when the main branch of this repo includes at least one MSVC flavored build.

### PR DESCRIPTION
**Pull request by wusatosi**
This PR adds MSVC CI support.

Could be improved after #76 is merged.

Conflict with unknown state PR #66 .

Closes #24 .

### CODE REVIEW

File: .github/workflows/ci_tests.yml
Comment (camio) [inline .github/workflows/ci_tests.yml:71]:
```suggestion
```

It isn't clear what is meant by "use include here". Also, it is unclear if #76 in its current form will be merged.

File: .github/workflows/ci_tests.yml
Comment (camio) [inline .github/workflows/ci_tests.yml:129]: Maybe something like this?
```suggestion
          cd /opt/beman.exemplar
          tree
          cd -
```

File: .github/workflows/ci_tests.yml
Comment (camio) [inline .github/workflows/ci_tests.yml:112]: Note that `--version` isn't a `cl.exe` flag. Maybe something like this? See [this reference page](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#operators).

```suggestion
          ${{ matrix.platform.cpp }} ${{ matrix.platform.cpp == 'cl' && '' || '--version' }}
          ${{ matrix.platform.c }}  ${{ matrix.platform.cpp == 'cl' && '' || '--version' }}
```

**Review by camio — COMMENTED** [review]
A couple nits on platform-independent syntax, but otherwise looks good.

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml:129]: that's a great idea!

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: I don't think this is working.

Also `--version` seem to work? (I know this shouldn't but it's working).
image

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml:71]: yeah sorry I should check my comment super well.

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: Switched to a GitHub action runner provided directory. Still got empty directories from tree command, I don't know if `tree` is just badly implemented under bash environment

Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: Agh, it's a quirk with powershell tree.

See: https://superuser.com/questions/1270040/windows-10-powershell-tree-command-how-to-show-files

Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: switch out to `ls -R`, and it is working. It's just very ugly...

image
image

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: done.

File: .github/workflows/ci_tests.yml
Comment (neatudarius) [inline .github/workflows/ci_tests.yml:135]: why `find` is not good?

File: .github/workflows/ci_tests.yml
Comment (neatudarius) [inline .github/workflows/ci_tests.yml:35]:
```suggestion
          - description: "Ubuntu GCC"
```
if it's in description?

File: .github/workflows/ci_tests.yml
Comment (neatudarius) [inline .github/workflows/ci_tests.yml:37]: Do we ever use C compiler? More a curiosity.

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml:135]: Because we need to execute this on windows runner for MSVC, where tree doesn't work with the /opt/.... Directory.

Should I include a comment here documenting this?

Comment (neatudarius) [reply .github/workflows/ci_tests.yml:135]: OK, got it, right.  I would add a command like `name: Build Release # Portable commands only`. thanks!

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml:37]: This is carried over from previous commits. 

We can evaluate this later

Comment (steve-downey) [reply .github/workflows/ci_tests.yml:37]: > Do we ever use C compiler? More a curiosity.

Probably not us deliberately, but Cmake will look for one anyway, and any packages we fetch might use it. When this goes wrong it's surprisingly difficult to track down.

Comment (neatudarius) [reply .github/workflows/ci_tests.yml:37]: Let's keep it.

File: .github/workflows/ci_tests.yml
Comment (camio) [reply .github/workflows/ci_tests.yml]: You can use the same trick I suggested in [this comment](https://github.com/bemanproject/exemplar/pull/80#discussion_r1870168036) to use the `/F` flag on windows only

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml:136]: Adopted.

File: .github/workflows/ci_tests.yml
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: Adopted.

**Review by camio — APPROVED** [review]
The `# Portable commands only` comments clutter up the code in my opinion, but I don't want to hold up merging for that :-D

**Comment by wusatosi** [comment]
> The `# Portable commands only` comments clutter up the code in my opinion, but I don't want to hold up merging for that :-D

Yeah I do agree it clutters up the code...

File: .github/workflows/ci_tests.yml
Comment (steve-downey) [inline .github/workflows/ci_tests.yml:144]: This install isn't going to do anything because it's installing to the same location as the Release install. Unless that's the intent?

Comment (wusatosi) [reply .github/workflows/ci_tests.yml:144]: Intention here is to test the install command.

Comment (steve-downey) [reply .github/workflows/ci_tests.yml:144]: It ends up skipping installation because it's already installed, which is a sort of test. Not critical for header only, but if there's any generated config or binaries, they shouldn't clobber each other.

**Comment by steve-downey** [comment]
Perhaps an explanatory general comment, once, reminding us to use the portable subset of commands. 
On the other hand, if it breaks in CI, a PR that doesn't use the portable subset will fail?

**Comment by wusatosi** [comment]
> Perhaps an explanatory general comment, once, reminding us to use the portable subset of commands. On the other hand, if it breaks in CI, a PR that doesn't use the portable subset will fail?

Ops I didn't see this before merging. I don't think this is fine. We can move this out in a later PR.

**Pull request merged**
